x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java

```diff
@@ -0,0 +1,320 @@
+        if (this.inferenceResults != null) {
+            // The supplier is generally not used after the results are set, but set the supplier to maintain equality when serializing
+            // & deserializing
+            this.inferenceResultsSupplier = new SetOnce<>(this.inferenceResults);
+        }
+    }
+
+    private SemanticQueryBuilder(
+        SemanticQueryBuilder other,
+        SetOnce<InferenceServiceResults> inferenceResultsSupplier,
+        InferenceServiceResults inferenceResults
+    ) {
+        this.fieldName = other.fieldName;
+        this.query = other.query;
+        this.boost = other.boost;
+        this.queryName = other.queryName;
+        this.inferenceResultsSupplier = inferenceResultsSupplier;
+        this.inferenceResults = inferenceResults;
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersions.SEMANTIC_TEXT_FIELD_ADDED;
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        if (inferenceResultsSupplier != null && inferenceResults == null) {
+            throw new IllegalStateException("Inference results supplier is set, but inference results is null. Missing a rewriteAndFetch?");
+        }
+        out.writeString(fieldName);
+        out.writeString(query);
+        out.writeOptionalNamedWriteable(inferenceResults);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.startObject(fieldName);
+        builder.field(QUERY_FIELD.getPreferredName(), query);
+        boostAndQueryNameToXContent(builder);
+        builder.endObject();
+        builder.endObject();
+    }
+
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
+        // We cannot fully rewrite the query to a NestedQueryBuilder here because that query builder validates that the path is
+        // registered as a nested object. This is not the case for semantic_text fields; the SemanticTextInferenceResultFieldMapper
+        // metafield mapper indexes inference results for the field using a "shadow" nested field mapping that is coordinated between
+        // this query and the mapper. This "shadow" mapping is not registered with the ES index mappings.
+        //
+        // Instead, we extract the inference results from the supplier in a serializable format and handle creation of the Lucene query
+        // in this class in doToQuery.
+        if (inferenceResults != null) {
+            return this;
+        }
+
+        if (inferenceResultsSupplier != null) {
+            InferenceServiceResults inferenceResults = inferenceResultsSupplier.get();
+            return inferenceResults != null ? new SemanticQueryBuilder(this, inferenceResultsSupplier, inferenceResults) : this;
```

Comment:
Fixed a subtle bug here. Previously, the code in this conditional set the inference results by modifying this instance. This caused the query rewrite loop to terminate one cycle early, which happened to work in this case because setting the inference results is the last thing we (currently) care about in the rewrite process.

However, as we add more steps to the rewrite process, this early termination would have become a problem.